protocols/ML01.md

```diff
@@ -0,0 +1,315 @@
+- The API has been generalised to not specifically reference training/models and instead refence the more general "jobs":
+  - Gives more flexibility to the implementation for data processing tasks that do not fit this pattern.
+  - Removes burden on client.
+  - Jobs referencing models are still posibile (see implementation notes)
+- Response format made consistent with BE01.
+- Spec now only uses GET and POST, as these are the request types with most widely available browser support.
+  - This allows the front end to not require a server component (this was a principle agreed during meetings).
+
+# Foreword: Required Reading
+
+The reader is assumed to be fully familiar with the contents of BE01.
+
+Reference is also made to the bundle files and report format of UI01.
+
+# Foreword: URL
+
+Within this document the URL `http://ml.endpoint/` will be used to indicate the base URL of the backend server's API endpoint.
+
+# Foreword: Matching notation
+
+The matching notation used in this specification is identical to that used in BE01.
+
+# Overview
+
+The server provides the ability to run data processing "jobs". These jobs can take parameters which the front end should collect from the end-user. The server then allows monitoring of the status of these jobs. The server accesses the backend as a normal client using a provided authentication token. It reads data files from the backend and writes output files back there. Generally the server will write one or more result data files and a single report file into a directory and mark the directory as bundle (opening the report file by default) (see UI01 for details of this).
+
+# Basic Response
+
+The basic reponse is identical to that used in BE01.
+
+# Authentication
+
+The client **MUST** send its authorisation token, as obtained from the OAuth endpoint, in the headers of each request as described in BE01.
+
+The server **SHOULD** verify the token by using it to access the current user details API (or some other authenticated API) on the backend.
+
+If the server cannot verify the identity of the user it **SHOULD** return http status 401 and error "not_authorised".
+
+A possible deviation from the specified behaviour would be for the server to only check token validity when required to access some backend resource. The is **not** recommended and an implementation that does such should ensure that it fully understands and addresses security consequences of doing so.
+
+# Listing Available Jobs
+
+In order to request the data processing tasks that the server is capable of, clients shoud request:
+
+```
+http://ml.endpoint/projects/<project_name>/job_types
+```
+
+If the given project does not exist the server **SHOULD** return error code 404 and error "project_not_found".
+
+If the user does not have accessed to the project an unauthorised response **SHOULD** be returned. (see above)
+
+If the project exists and the user is authorised the server's unwrapped response **MUST** match
+
+```javascript
```

Comment:
These responses/requests don't follow the JSON schema, which we had previously agreed to use.

Reply:
assuming this means jsonapi - maybe I misunderstand this specification but it seems to be that the other protocols don't either? Responses "data" attributes don't seem to be valid  “Resource objects”  as the spec requires? A confusion about exactly how and what features of jsonapi are being used  led me to develop the simplest response format that I felt met the needs. I don't foresee massive difficulty in re-mapping the response attributes to match it pending clarification of exactly what features are to be used - personally jsonapi seems a bit overbearing.

Reply:
I don't really have an opinion on whether the jsonapi is a good idea or not, it's just something that we'd agreed to use. From http://jsonapi.org/format/ the "data" attribute is described:

"
Top Level

A JSON object MUST be at the root of every JSON API request and response containing data. This object defines a document’s “top level”.

A document MUST contain at least one of the following top-level members:

    data: the document’s “primary data”
    errors: an array of error objects
    meta: a meta object that contains non-standard meta-information.
"

I was trying to follow the schema when I wrote the Swagger, but it's certainly possible I made mistakes.

Reply:
I was referring to 
```
A document MUST contain at least one of the following top-level members:
...
data: the document’s “primary data”
```

```
Primary data MUST be either:

    a single resource object, a single resource identifier object, or null, for requests that target single resources
    an array of resource objects, an array of resource identifier objects, or an empty array ([]), for requests that target resource collections
```
and
```
A resource object MUST contain at least the following top-level members:

    id
    type
```